src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -87,115 +109,119 @@ EmbeddedAssemblies::zip_load_entries (int fd, const char *apk_name, monodroid_sh
 		if (compression_method != 0)
 			continue;
 
-		if (strncmp (prefix, file_name, prefix_len) != 0)
+		if (strncmp (prefix, entry_name.get (), prefix_len) != 0)
 			continue;
 
 #if defined (NET6)
 		if (application_config.have_runtime_config_blob && !runtime_config_blob_found) {
-			if (utils.ends_with (file_name, SharedConstants::RUNTIME_CONFIG_BLOB_NAME)) {
+			if (utils.ends_with (entry_name, SharedConstants::RUNTIME_CONFIG_BLOB_NAME)) {
 				runtime_config_blob_found = true;
-				runtime_config_blob_mmap = md_mmap_apk_file (fd, data_offset, file_size, file_name, apk_name);
+				runtime_config_blob_mmap = md_mmap_apk_file (fd, data_offset, file_size, entry_name.get (), apk_name);
 				continue;
 			}
 		}
 #endif // def NET6
 
 		// assemblies must be 4-byte aligned, or Bad Things happen
 		if ((data_offset & 0x3) != 0) {
-			log_fatal (LOG_ASSEMBLY, "Assembly '%s' is located at bad offset %lu within the .apk\n", file_name, data_offset);
+			log_fatal (LOG_ASSEMBLY, "Assembly '%s' is located at bad offset %lu within the .apk\n", entry_name.get (), data_offset);
 			log_fatal (LOG_ASSEMBLY, "You MUST run `zipalign` on %s\n", strrchr (apk_name, '/') + 1);
 			exit (FATAL_EXIT_MISSING_ZIPALIGN);
 		}
 
-		bool entry_is_overridden = !should_register (strrchr (file_name, '/') + 1);
+		const char *last_slash = utils.find_last (entry_name, '/');
+		bool entry_is_overridden = last_slash == nullptr ? false : !should_register (last_slash + 1);
 
-		if ((utils.ends_with (file_name, ".pdb") || utils.ends_with (file_name, ".mdb")) &&
+		if ((utils.ends_with (entry_name, ".pdb") || utils.ends_with (entry_name, ".mdb")) &&
 				register_debug_symbols &&
 				!entry_is_overridden &&
-				bundled_assemblies != nullptr) {
-			md_mmap_info map_info = md_mmap_apk_file(fd, data_offset, file_size, file_name, apk_name);
-			if (register_debug_symbols_for_assembly (file_name, (bundled_assemblies) [bundled_assemblies_count - 1], (const mono_byte*)map_info.area, static_cast<int>(file_size)))
+				bundled_assembly_index >= 1) {
+			md_mmap_info map_info = md_mmap_apk_file(fd, data_offset, file_size, entry_name.get (), apk_name);
+			if (register_debug_symbols_for_assembly (entry_name, (bundled_assemblies) [bundled_assembly_index - 1], (const mono_byte*)map_info.area, static_cast<int>(file_size)))
 				continue;
 		}
 
 #if !defined(NET6)
-		if (utils.ends_with (file_name, ".config") && bundled_assemblies != nullptr) {
-			char *assembly_name = strdup (basename (file_name));
+		if (utils.ends_with (entry_name, ".config") && !bundled_assemblies.empty ()) {
+			char *assembly_name = strdup (basename (entry_name.get ()));
```

Comment:
I look at this, wonder "why `char*` and not `std::string`", look at the source to `mono_register_config_for_assembly()`, and see that mono doesn't duplicate the string, but just uses it as-is.  Meaning we *cannot* free this value, and thus we shouldn't use `std::string`.

Which makes me wonder if there's a more "type declarative" approach here, to say "we explicitly relinquish ownership of this value to mono"?

…and I think there is/could be: `std::unique_ptr::release`:

```cpp
c_unique_ptr<char[]> assembly_name {strdup (basename (entry_name.get ()))};
*strrchr (assembly_name.get (), '.') = '\0';
// …
mono_register_config_for_assembly (assembly_name.release (), (const char *) map_info.area);
```

"Pro": this is exception safe.  (Not that we expect or should otherwise be using exceptions, as who knows what all that'll break.  So possibly a con?)

"Con": perhaps more overhead because of the use of `c_unique_ptr`, though I'm not sure how much this would actually impact anything.

Reply:
Your observations are valid, and normally I'd agree that we should use both `std::string` and `std::unique_ptr` but they do carry an overhead, and in this code we're trying to gain as much performance as possible, thus `char*`. Regarding exceptions - I'd like to use them for "real" RAII but enabling them without linking `libc++` and `libc++abi` is not a quick task. Exceptions require a bit of infrastructure from both of these libraries, code we'd either have to implement or port from `libc++` sources. I looked at it briefly and it's something that would require a bit of time to implement.